pyartcd/pyartcd/util.py

```diff
@@ -267,3 +275,191 @@ async def is_build_permitted(version: str, data_path: str = constants.OCP_BUILD_
+    operator_registry = get_release_image_pullspec(f"{quay_url}:{from_release_tag}", "operator-registry")
+    binaries = ['opm']
+    platforms = ['linux']
+    if arch == 'x86_64':  # For x86_64, we have binaries for macOS and Windows
+        binaries += ['darwin-amd64-opm', 'windows-amd64-opm']
+        platforms += ['mac', 'windows']
+    path_args = []
+    for binary in binaries:
+        path_args.append(f'--path=/usr/bin/registry/{binary}:{CLIENT_MIRROR_DIR}')
+    extract_release_binary(operator_registry, path_args)
+    # Compress binaries into tar.gz files and calculate sha256 digests
+    os.chdir(CLIENT_MIRROR_DIR)
+    for idx, binary in enumerate(binaries):
+        platform = platforms[idx]
+        os.chmod(binary, 0o755)
+        with tarfile.open(f"opm-{platform}-{release_name}.tar.gz", "w:gz") as tar:  # archive file
+            tar.add(binary)
+        os.remove(binary)  # remove oc-mirror
+        os.symlink(f'opm-{platform}-{release_name}.tar.gz', f'opm-{platform}.tar.gz')  # create symlink
+        with open(binary, 'rb') as f:  # calc shasum
+            shasum = hashlib.sha256(f.read()).hexdigest()
+        with open("sha256sum.txt", 'a') as f:  # write shasum to sha256sum.txt
+            f.write(shasum)
+
+    print_dir_tree(CLIENT_MIRROR_DIR)  # print dir tree
+    print_file_content(f"{CLIENT_MIRROR_DIR}/sha256sum.txt")  # print sha256sum.txt
+
+    # Publish the clients to our S3 bucket.
+    subprocess.run(f"aws s3 sync --no-progress --exact-timestamps {BASE_TO_MIRROR_DIR}/ s3://art-srv-enterprise/pub/openshift-v4/", shell=True, check=True)
+
+
+def stagePublishMultiClient(working_dir, from_release_tag, release_name, client_type):
+    subprocess.run(f"docker login -u openshift-release-dev+art_quay_dev -p {os.environ['PASSWORD']} quay.io")
+    # Anything under this directory will be sync'd to the mirror
+    BASE_TO_MIRROR_DIR = f"{working_dir}/to_mirror/openshift-v4"
+    shutil.rmtree(BASE_TO_MIRROR_DIR, ignore_errors=True)
+    RELEASE_MIRROR_DIR = f"{BASE_TO_MIRROR_DIR}/multi/clients/{client_type}/{release_name}"
+
+    for goArch in goArches:
+        if goArch == "multi":
+            continue
+        # From the newly built release, extract the client tools into the workspace following the directory structure
+        # we expect to publish to mirror
+        CLIENT_MIRROR_DIR = f"{RELEASE_MIRROR_DIR}/{goArch}"
+        os.makedirs(CLIENT_MIRROR_DIR)
+        # extract release clients tools
+        extract_release_client_tools(f"{constants.QUAY_URL}:{from_release_tag}", f"--to={CLIENT_MIRROR_DIR}", goArch)
+        # create symlink for clients
+        create_symlink(CLIENT_MIRROR_DIR, True, True)
+
+    # Create a master sha256sum.txt including the sha256sum.txt files from all subarches
+    # This is the file we will sign -- trust is transitive to the subarches
+    subprocess.run(f"cd {RELEASE_MIRROR_DIR};sha256sum */sha256sum.txt > {RELEASE_MIRROR_DIR}/sha256sum.txt", shell=True, check=True)
```

Comment:
This function is called within a `for` loop; making this commands async would let us run things in parallel and save execution time. Why is `subprocess.run` needed?

Reply:
I copy the method from Jenkinsfile, we can optimize later, the shasum method can implement in python method

Reply:
Unless this is urgent, why not doing it right away? We'll almost certainly forget to address this later